### ISSUE TITLE
Periodic approximations

### ISSUE THREAD

**Issue by ivan-pi** [treverhines/RBF#12]
Hello,

for some numerical benchmarks it would be useful to have a version of the `weights` and `weight_matrix` functions that would take into account the periodicity of the domain. Given a point where we seek an approximation and a set of centers (that have already been selected taking periodicity into account) it is only necessary to add a correction during the calculation of the distance matrix (for both the RBF  part and shifted monomials). For example, if the domain is periodic in x-direction:

```
if (periodic_x):
  dx = x(j) - x(i)
  if (dx >   x_size * 0.5): dx = dx - x_size
  if (dx <= -x_size * 0.5): dx = dx + x_size
end if
```

The `dx` (and `dy`) are then used to calculate the radial distance that goes into the RBF. I had a look at the internals of the library  and identified line 317 in `pde/basis.py` as the place where this is performed on symbolic level. Do you have any suggestion how to go about this? Is the easiest way just to create 9 copies of the nodes (e.g. for bi-periodic in 2d), and then loop through the nodes calling the `weights` function, and place the weights into a sparse matrix?

I also have a question whether there is any reason when you fill the RBF-FD matrix, you do not shift the monomial part by the centers of the approximation?

Thank you for sharing your work.

Ivan

**Comment by treverhines** [comment]
Hi Ivan,

To solve a PDE over a periodic domain with RBF-FD, you need to rewrite your PDE to be in terms of cartesian coordinates. Then you build your differentiation matrices using `weight_matrix` for the cartesian coordinates. 

For example, suppose we want to build a matrix to differentiate 1-dimensional periodic functions with period 2*pi. We first convert the dependent variable, t, to the cartesian variables x and y as 

x=cos(t)
y=sin(t)

Differentiation with respect to t can be rewritten in terms of differentiation with respect to x and y as

d/dt = d/dx * dx/dt  + d/dy * dy/dt = d/dy * x - d/dx * y    

We can then feed the rewritten differential operator into `weight_matrix` by giving it the arguments `diffs=[(0, 1), (1, 0)]`, and `coeffs=[x, -y]`.  

Here is some code to make this example complete:
```python
import numpy as np                                                              
import matplotlib.pyplot as plt                                                 
from rbf.pde.fd import weight_matrix                                            
                                                                                
t = np.linspace(0.0, 2*np.pi, 21)[:-1]                                          
x = np.cos(t)                                                                   
y = np.sin(t)                                                                   
xy = np.array([x, y]).T                                                         
D = weight_matrix(xy, xy, 5, diffs=[(0, 1), (1, 0)], coeffs=[x, -y])            
                                                                                
example_function = np.sin(t) + np.cos(2*t)                                      
true_derivative =  np.cos(t) - 2*np.sin(2*t)                                    
approx_derivative = D.dot(example_function)                                     
                                                                                
plt.figure(1)                                                                   
p = plt.imshow(D.A)                                                             
plt.colorbar(p)                                                                 
plt.title('differentiation matrix')                                             
                                                                                
plt.figure(2)                                                                   
plt.plot(t, example_function, 'C0o-', label='sin($\\theta$) + cos($2\\theta$)') 
plt.plot(t, true_derivative, 'C1o-', label='true derivative')                   
plt.plot(t, approx_derivative, 'C2o--', label='approx. derivative')             
plt.xlabel('t')                                                                 
plt.legend()                                                                    
plt.show()
```
You should then get the following figures, where you can see that the differentiation matrix is indeed periodic:

![Figure_2](https://user-images.githubusercontent.com/7061940/64479508-23535f00-d186-11e9-990e-ce1347f1204f.png)

![Figure_1](https://user-images.githubusercontent.com/7061940/64479505-16367000-d186-11e9-9f4e-f9384fbd6b2c.png)

Regarding shifting the monomials: In theory, the RBF-FD weights will be the same regardless of whether the monomials are centered at the stencil. In reality, if the domain is poorly scaled and/or high order monomials are used, then the RBF-FD weights can be influenced by numerical rounding errors. So it is a good idea to center the monomials, and I will make that fix.    

I hope this helps.

-Trever

**Comment by ivan-pi** [comment]
Hello Trever,

thank you for the very neat example. I am impressed by the design of the weight_matrix function that allows such thoughtful use of the chain rule via the `coeffs` variable. However, I still think this approach has some deficiencies. I have prepared a small 2D demonstration to show what I have in mind.

Let's consider a simple 2D example by mapping the periodic space `(t,r)` on to a torus `(x,y,z)` like suggested in the book by Fornberg & Flyer and or section 4.3 of [this paper](http://citeseerx.ist.psu.edu/viewdoc/download?doi=10.1.1.716.9284&rep=rep1&type=pdf).
I will try to approximate the gradient d/dt + d/dr of a function:

```Python
import numpy as np
import matplotlib.pyplot as plt
from rbf.pde.fd import weight_matrix

# create a regular grid
st, dt = np.linspace(0,2*np.pi,32,endpoint=False,retstep=True)
sr, dr = np.linspace(0,2*np.pi,32,endpoint=False,retstep=True)
t, r = np.meshgrid(st,sr)
t, r = t.flatten(), r.flatten()

# conformal mapping to the surface of a torus
f = 1./(np.sqrt(2.) - np.cos(r))
x = f*np.cos(t)
y = f*np.sin(t)
z = f*np.sin(r)

# fig3d, ax3d = plot_torus(x,y,z)
# plt.show()

xyz = np.stack((x,y,z),axis=-1)

# partial derivatives used by the chain rule (obtained using sympy)
dx_dt = -y
dy_dt = x
dz_dt = np.zeros_like(x)
dx_dr = -z*x
dy_dr = -z*y
dz_dr = np.sqrt(2)*np.sqrt(x**2 + y**2) - 1. - z**2

diffs=[[1,0,0],[0,1,0],[0,0,1]]
coeffs=[dx_dt + dx_dr, dy_dt + dy_dr, dz_dt + dz_dr]
D = weight_matrix(xyz,xyz,21,diffs=diffs,coeffs=coeffs,phi='phs3').tocsr()


example_function = np.sin(t)*np.cos(2*r)
approx_derivative = D.dot(example_function)

fine_grid = np.linspace(0,2*np.pi,100,endpoint=False)
tt,rr = np.meshgrid(fine_grid,fine_grid)
true_derivative = np.cos(2*rr)*np.cos(tt) - np.sin(tt)*2*np.sin(2*rr)
vmin, vmax = true_derivative.min(), true_derivative.max()

plt.figure(1)
cb = plt.scatter(t,r,s=20,c=example_function,vmin=-1.0,vmax=1.0)
plt.title('Example function')
plt.xlabel(r'$t$')
plt.ylabel(r'$r$')
plt.colorbar(cb)
plt.axis('equal')
plt.tight_layout()

plt.figure(2)
cntr1 = plt.pcolormesh(tt,rr,true_derivative,vmin=vmin,vmax=vmax)
plt.scatter(t,r,s=20,c=approx_derivative,edgecolor='k',vmin=vmin,vmax=vmax)
plt.title("RBF-FD approximated gradient")
plt.xlabel(r'$t$')
plt.ylabel(r'$r$')
plt.colorbar(cntr1)
plt.axis('equal')
plt.tight_layout()

plt.show()
```
The output are the following figures:
![function](https://user-images.githubusercontent.com/21085643/64492525-f16cf600-d274-11e9-83c9-154cfafc12ac.png)
![gradient](https://user-images.githubusercontent.com/21085643/64492526-f5991380-d274-11e9-9292-6e70a0e44a9b.png)
The RBF-FD gradient is shown as the scattered points.

Looking at the sparsity plot of the resulting matrix it is easy to observe that
it is not symmetric, like we would expect, if every node had the same local stencil shape.
This is due to the fact, that the torus mapping is only locally distance preserving.
Instead of getting the same FD stencil for each node, we get different stencils
depending on the nearest neighbors on the torus (for example on the inside of the torus the points are spaced closer together). To fix this I have created a separate version of the `weight_matrix` function that creates a periodic tiling of nodes, and then performs the nearest neighbor searches in the center tile of the larger domain (see my [pull request](https://github.com/treverhines/RBF/pull/13)). The code to draw the sparsity plots:

```Python
from rbf.pde.fd import periodic_weight_matrix

bbox = np.array([[0.,0.],[2*np.pi,2.*np.pi]])
tr = np.stack((t,r),axis=-1)
D_tr = periodic_weight_matrix(bbox,tr,tr,21,diffs=[[1,0],[0,1]],phi='phs3').tocsr()

fig, (ax1,ax2) = plt.subplots(nrows=1,ncols=2,sharey=True)

ax1.spy(D,color='C0')
ax1.set_xlabel('Conformal mapping')
ax2.spy(D_tr,color='C1')
ax2.set_xlabel('Direct evaluation')

plt.show()
```
![sparsity_patterns](https://user-images.githubusercontent.com/21085643/64492747-a6081700-d277-11e9-9706-e5fc8f9922fc.png)

Another issue with the mapping method is that for RBF-FD approximations augmented with monomial terms, the order parameter has a different meaning (e.g. in 1D I would like to use 1, x, x^2 and three PHS terms on a 6-node stencil, but with the "2D" differentiation matrix based on the circle mapping 6 nodes are only able to support the linear monomial terms).

Hopefully, this post is explanatory enough.

Ivan

**Comment by treverhines** [comment]
Hi Ivan,

Thank you for your contribution. However, I maintain that the appropriate way to deal with a periodic domain is to map the PDE to cartesian coordinates.  

I disagree with your argument that we would expect the differentiation matrix to be symmetric. RBF-FD differentiation matrices are generally not symmetric due to the fact that the nodes are generally not regularly spaced.   

I see your point about the monomials not being what you want with the mapping method. For the case of a 1-D periodic domain and `order=2`, the monomials would be {1, t, t^2} with the tiling method and {1, sin(t), cos(t), sin^2(t), cos^2(t), sin(t)*cos(t)} with the mapping method. It is not immediately clear to me which set of basis functions would be better, although I am inclined to prefer the latter because they are periodic.  

If you still want to go the tiling route, then I think the solution is much simpler than the code in your pull request. You would first tile the domain, then give `weight_matrix` the untiled and tiled nodes, then replace the column indices in the output COO matrix with their remainder when dividing by the number of untiled nodes. There is no need to rewrite the `weight_matrix` function. For example, the differentiation matrix from my example above can be created using the tiling method with the following

```python
import numpy as np                                                              
import matplotlib.pyplot as plt                                                 
from rbf.pde.fd import weight_matrix                                            
from scipy.sparse import coo_matrix                                             
                                                                                
n = 20                                                                          
t = np.linspace(0.0, 2*np.pi, n + 1)[:-1, None]                              
t_tiled = np.vstack((t - 2*np.pi, t, t + 2*np.pi))                              
                                                                                
D = weight_matrix(t, t_tiled, 5, diffs=(1,))                                    
D = coo_matrix((D.data, (D.row, D.col%n)), shape=(n, n))                        
                                                                                
p = plt.imshow(D.A)                                                             
plt.show()  
```

![Figure_1](https://user-images.githubusercontent.com/7061940/64645381-ca343700-d3e2-11e9-9a44-9a1a48818925.png)

Since I do not prefer the tiling strategy and also since the code for the tiling strategy is straight forward, I do not think it warrants its own function in the repository.

**Comment by ivan-pi** [comment]
Hi Trever,

after seeing your solution with the tiling strategy I can agree that a separate function is completely unnecessary. Coming from a Fortran and C++ background I tend to forget how easily things can be in Python even on the user application level and I did not realize the remainder trick could be applied directly to the sparse matrix triplets. Moreover, I am still able to apply my tiling function.

You are right that my concerns are restricted to regular node sets. The reason I came up with this issue is because I would like to benchmark some strategies for solving the advection equation comparing RBF-FD and WLS methods on both regular grids and scattered node sets. At least for regular grids, having a symmetric stencil can help guarantee the matrix has only real eigenvalues meaning one can safely use some of the Runge-Kutta methods without any additional hyper-viscosity terms. Concerning which set of basis functions is preferable I suppose is problem dependent (e.g. if we plan to use an explicit or implicit time-stepping method). Perhaps by looking at the eigenvalues of the resulting matrices could provide some further insight how the discrete operators differ, even though the physical and mapped (Cartesian) domains are algebraically equivalent.

Thanks for the support!
-Ivan
